asdf/compression.py

```diff
@@ -30,118 +34,232 @@ def validate(compression):
-            self._buffer[self._pos:self._pos + len(data)] = data
-            self._pos += len(data)
-            return b''
-        else:
-            offset = self._size - self._pos
-            self._buffer[self._pos:] = data[:offset]
-            data = data[offset:]
-            self._size = 0
-            self._pos = 0
-            output = self._api.decompress(self._buffer)
-            self._buffer = b''
-            return output + self.decompress(data)
-
-
-def _get_decoder(compression):
-    if compression == 'zlib':
-        try:
-            import zlib
-        except ImportError:
-            raise ImportError(
-                "Your Python does not have the zlib library, "
-                "therefore the compressed block in this ASDF file "
-                "can not be decompressed.")
-        return zlib.decompressobj()
-    elif compression == 'bzp2':
-        try:
-            import bz2
-        except ImportError:
-            raise ImportError(
-                "Your Python does not have the bz2 library, "
-                "therefore the compressed block in this ASDF file "
-                "can not be decompressed.")
-        return bz2.BZ2Decompressor()
-    elif compression == 'lz4':
         try:
             import lz4.block
         except ImportError:
             raise ImportError(
                 "lz4 library in not installed in your Python environment, "
                 "therefore the compressed block in this ASDF file "
                 "can not be decompressed.")
-        return Lz4Decompressor(lz4.block)
-    else:
-        raise ValueError(
-            "Unknown compression type: '{0}'".format(compression))
 
+        self._api = lz4.block
 
-def _get_encoder(compression):
-    if compression == 'zlib':
-        try:
-            import zlib
-        except ImportError:
-            raise ImportError(
-                "Your Python does not have the zlib library, "
-                "therefore the block in this ASDF file "
-                "can not be compressed.")
-        return zlib.compressobj()
-    elif compression == 'bzp2':
-        try:
-            import bz2
-        except ImportError:
-            raise ImportError(
-                "Your Python does not have the bz2 library, "
-                "therefore the block in this ASDF file "
-                "can not be compressed.")
-        return bz2.BZ2Compressor()
-    elif compression == 'lz4':
-        try:
-            import lz4.block
-        except ImportError:
-            raise ImportError(
-                "lz4 library in not installed in your Python environment, "
-                "therefore the block in this ASDF file "
-                "can not be compressed.")
-        return Lz4Compressor(lz4.block)
+        # compression
+        self._mode= compression_kwargs.get('mode','default')
+        self._compression_block_size = compression_kwargs.get('compression_block_size',1<<22)
+
+        # decompression
+        self._size = 0
```

Comment:
Can these variables be local to the decompress method now that it's handling the entire input in one call?